docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+* Type annotations
+
+In Elisp users are not required to provide type annotations to their
+code.  While at many places the types can be inferred there are
+places, especially in user-defined functions, where we can not infer
+the correct type.
+
+Therefore Elsa provides users with the option to annotate their
+function definitions.  The annotations are placed in the body of the
+function inside a =declare= form:
+
+#+BEGIN_SRC elisp
+(defun add-one (x)
+  (declare (elsa (int) int))
+  (1+ x))
+#+END_SRC
+
+The =declare= form starts with =elsa= followed by one or two values.  If
+two are provided, first must be a list which is the list of input
+types and the second is the return type.
+
+If only one is provided it is taken to be the return type and the
+function must have no input arguments.
+
+Here are general guidelines on how the types are constructed.
+
+** Built-in types
+
+For built-in types with test predicates, drop the =p= or =-p= suffix to
+get the type:
+
+- =stringp= → =string=
+- =integerp= → =integer= (=int= is also accepted)
+- =markerp= → =marker=
+- =hash-table-p= → =hash-table=
+
+Some additional special types:
+
+- =t= stands for =t= and is always true
+- =nil= stands for =nil= and is always false
+- =bool= is a combination of explicitly =t= or =nil=
+
+** Nullable types
+
+By default types do not accept =nil= values.  This helps preventing
+errors where you would pass =nil= to a function expecting a value.  To
+mark a type nullable you can combine it with =nil= type using the [[id:5a21a68a-4df1-4d44-a854-1d9700858a1a][or]]
+combinator: =(or string nil)= is a type which takes any string but also
```

Comment:
I don't think you can combine `*` and `?` in the same type because what is before `*` is an abstract name that conveys no meaning by itself. My opinion is that `a*?` and `a?*` should trigger an error.

Reply:
If you know Haskell I can think of this as an analogy of `Maybe a`, that would be in our syntax `a*?` (since `a*` is the generic type constructor).  So basically `?` is short for `(or nil <arg>)`, similar as one can write `(list a*)`

`a?*` probably makes no sense other than `a?` would be the placeholder and that shouldn't be allowed indeed.

We can also put some things to the front, (e)lisp has a tradition with the `&` prefix.

Or we can go entirely different way and for example use strings or quoted symbols for generic names, so that it would be `"foo"` or (list "foo")` instead of `foo*`.